### PR TITLE
[FEATURE] Add IMU Sensor

### PR DESCRIPTION
**Pull request by Milotrince**
## Description
Add IMU sensor.

## Related Issue
https://github.com/Genesis-Embodied-AI/Genesis/issues/1513

## Motivation and Context
See related issue

## How Has This Been / Can This Be Tested?
Added test in `test_sensors.py`

## Checklist:
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.

### CODE REVIEW

File: genesis/engine/entities/rigid_entity/rigid_entity.py [hunk 1]
```diff
@@ -1730,27 +1730,6 @@ def get_links_ang(self, links_idx_local=None, envs_idx=None, *, unsafe=False):
...
-            The indices of the environments. If None, all environments will be considered. Defaults to None.
-
-        Returns
-        -------
-        acc : torch.Tensor, shape (n_links, 3) or (n_envs, n_links, 3)
-            The accelerometer data of IMUs rigidly attached of the specified entity's links.
-        """
-        links_idx = self._get_idx(links_idx_local, self.n_links, self._link_start, unsafe=True)
-        return self._solver.get_links_acc(links_idx, envs_idx, mimick_imu=True, unsafe=unsafe)
-
```
Comment (Milotrince) [inline genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: I could leave this here and add a deprecation warning instead, although imo it will be cleaner to remove.

File: genesis/sensors/imu.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        ret = []
+        if self._options.return_accelerometer:
+            ret.append(self._cache[envs_idx, self._cache_idx, 0, :].squeeze())
+        if self._options.return_gyroscope:
+            ret.append(self._cache[envs_idx, self._cache_idx, 1, :].squeeze())
+        return tuple(ret)
+
+
+@register_sensor(IMU)
+class IMUOptions(SensorOptions):
```
Comment (Milotrince) [inline genesis/sensors/imu.py:97]: should the options be moved to options/ folder?

File: genesis/engine/entities/rigid_entity/rigid_entity.py (hunk 1 above)
Comment (Kashu7100) [reply genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: I prefer keeping this raw API. We don't want to instantiate Sensor every time if there is a "simple" way to access those.

File: genesis/engine/solvers/rigid/rigid_solver_decomp.py [hunk 3]
```diff
@@ -1984,18 +1984,16 @@ def get_links_ang(self, links_idx=None, envs_idx=None, *, unsafe=False):
         tensor = ti_field_to_torch(self.links_state.cd_ang, envs_idx, links_idx, transpose=True, unsafe=unsafe)
         return tensor.squeeze(0) if self.n_envs == 0 else tensor
 
-    def get_links_acc(self, links_idx=None, envs_idx=None, *, mimick_imu=False, unsafe=False):
```
Comment (Kashu7100) [inline genesis/engine/solvers/rigid/rigid_solver_decomp.py:1987]: I would also prefer to keep this API as well. Also "mimick" -> "mimic"?

File: genesis/sensors/imu.py (hunk 2 above)
Comment (Kashu7100) [reply genesis/sensors/imu.py:97]: Yes, this needs to go under options

File: genesis/engine/entities/rigid_entity/rigid_entity.py (hunk 1 above)
Comment (Milotrince) [reply genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: but adding IMU is simple, whereas the existence of this function is unclear

Comment (Kashu7100) [reply genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: Why? I think it is beneficial to have data accessor interface in Entity (use for GT signal for reward etc.)

**Comment by Kashu7100** [comment]
I just realized but this PR needs to go to original Genesis (instead of fork)?

File: genesis/engine/entities/rigid_entity/rigid_entity.py (hunk 1 above)
Comment (Milotrince) [reply genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: but you can get GT from sensor too?

**Comment by Milotrince** [comment]
> I just realized but this PR needs to go to original Genesis (instead of fork)?

I am aware. This is because I wanted to do "stacked" PRs but i cant do that on genesis repo because my branch doesnt exist on the original genesis

File: genesis/sensors/imu.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    inv_transform_by_trans_quat,
+    transform_quat_by_quat,
+)
+
+from .base_sensor import Sensor
+
+
+@ti.data_oriented
+class IMU(Sensor):
+    CACHE_DTYPE = torch.float32
```
Comment (Milotrince) [inline genesis/sensors/imu.py]: gs_float

File: genesis/sensors/imu.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            )
+            # acc/ang shape: (B, n_links, 3)
+            local_acc = inv_transform_by_trans_quat(acc, self._shared_metadata["offsets_pos"], offset_quats)
+            local_ang = inv_transform_by_trans_quat(ang, self._shared_metadata["offsets_pos"], offset_quats)
+
+            local_acc = local_acc - gravity.unsqueeze(1).repeat(1, local_acc.shape[1], 1)
+
+            # cache shape: (B, n_links, 2, 3)
+            self._cache.copy_(torch.cat([local_acc.unsqueeze(2), local_ang.unsqueeze(2)], dim=2))
+            self._set_cache_updated()
```
Comment (Milotrince) [inline genesis/sensors/imu.py]: have a `read_all`/`update_cache` function and `read` should be implemented in base sensor

File: genesis/sensors/imu.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    transform_quat_by_quat,
+)
+
+from .base_sensor import Sensor
+
+
+@ti.data_oriented
+class IMU(Sensor):
+    CACHE_DTYPE = torch.float32
+    CACHE_SHAPE = (2, 3)  # linear_acceleration, angular_velocity
```
Comment (Milotrince) [inline genesis/sensors/imu.py]: use flattened 6

File: genesis/sensors/imu.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        if envs_idx is None:
+            envs_idx = 0 if self._solver.n_envs == 0 else np.arange(self._cache.shape[0])
+
+        ret = []
+        if self._options.return_accelerometer:
+            ret.append(self._cache[envs_idx, self._cache_idx, 0, :].squeeze())
+        if self._options.return_gyroscope:
+            ret.append(self._cache[envs_idx, self._cache_idx, 1, :].squeeze())
+        return tuple(ret)
```
Comment (Milotrince) [inline genesis/sensors/imu.py]: return dict!

Comment (Milotrince) [reply genesis/sensors/imu.py]: base sensor -> read will use cache, return type can be a tensor or dict
get_data_structure -> dict of tensor shapes OR tensor shape

File: genesis/engine/entities/rigid_entity/rigid_entity.py (hunk 1 above)
Comment (Kashu7100) [reply genesis/engine/entities/rigid_entity/rigid_entity.py:1753]: But why need to delete?
